scalyr_agent/builtin_monitors/linux_process_metrics.py

```diff
@@ -619,163 +732,214 @@ class ProcessMonitor(ScalyrMonitor):
+        # a) cumulative metrics - only the delta of the last 2 recorded values is used (eg cpu cycles)
+        # b) absolute metrics - the last absolute value is used
+
+        running_pids_set = set(running_pids)
+
+        for pid, process_metrics in self.metrics_history.items():
+            for _metric, _metric_values in process_metrics.items():
+                if not self.running_total_metrics.get(_metric):
+                    self.running_total_metrics[_metric] = 0
+                if _metric.is_cumulative:
+                    if pid in running_pids_set:
+                        if len(_metric_values) <= 1:
+                            self.running_total_metrics[_metric] += (_metric_values[-1] if _metric_values else 0)
+                        else:
+                            self.running_total_metrics[_metric] += (_metric_values[-1] - _metric_values[-2])
+                else:
+                    # absolute metric - accumulate the last reported value
+                    self.running_total_metrics[_metric] += _metric_values[-1]
+
+    def _remove_dead_processes(self, running_pids):
+        # once the _calculate_running_total has calculated the running total of the
+        # metrics for the current epoch, remove the entries for the process id in the history
+        # NOTE: the removal of the contributions (if applicable) should already be done so
+        # removing the entry from the history is safe.
+
+        all_pids = self.metrics_history.keys()
+        for _pid_to_remove in list(set(all_pids) - set(running_pids)):
+            # for all the absolute metrics, decrease the count that the dead processes accounted for
+            del self.metrics_history[_pid_to_remove]
+
+        # if no processes are running, there is no reason to report the running metric data
+        if not running_pids:
+            self.running_total_metrics = {}
 
     def gather_sample(self):
-        """Collect the per-process metrics for the monitored process.
+        """Collect the per-process tracker for the monitored process(es).
 
-        If the process is no longer running, then attempts to match a new one.
+        For multiple processes, there are a few cases we need to poll for:
+
+        For `n` processes, if there are < n running processes at any given point,
+        we should see if there are any pids or matching expression that gives pids
+        that is currently not running, if so, get its tracker and run it.
+        It is possible that one or more processes ran its course, but we can't assume that
+        and we should keep polling for it.
         """
-        if self.__pid is not None and not self.__is_running():
-            self.__set_process(None)
 
-        if self.__pid is None:
-            self.__set_process(self.__select_process())
+        trackers = []
+        for _pid in list(self.__select_processes()):
+            trackers.append(ProcessTracker(_pid, self._logger, self.__id))
```

Comment:
We shouldn't create a new ProcessTracker every time through the gather sample if we already had one for the process last time through it.

Reply:
maintaining a state of the older ProcessTracker and calling collect on them has some unwanted side effects, also maintaining the state consumes some extra memory. I didn't think there was a lot of overhead.

Reply:
What unwanted side effects?  It is not the overhead I am worried about in terms of reusing the ProcessTracker.  Rather, it's about the naturalness of the abstraction and doing things in a common way.  If something is called ProcessTracker, than I expect it to track that process over a period of time.  Additionally, if we know we want to track a process, it just feels natural to me that we should create something to track it over its lifetime.. and that object only is let go once the process dies.

Also, there is an efficiency thing as well.  The stat readers will only open up their corresponding /proc/pid/XXX file once when the reader is created.  After that, everytime it wants to read the file, it seeks to zero and reads the contents.  The Linux operating system is suppose to like this better.